### PR TITLE
Verify token signature in constant time, Run cheap checks first in token validation process

### PR DESCRIPTION
**Pull request by leonklingele**
See commit messages for details

### CODE REVIEW

**Comment by leonklingele** [comment]
Please note that the second commit slightly changes (fixes?) behavior:
Before, the nonce would be updated in the DB (`db.exec("UPDATE nonces..`) even when the token contained an invalid signature. This is not longer the case and the call to the DB is only done when the token has been completely and successfully validated.

File: src/invidious/helpers/tokens.cr [hunk 1]
```diff
@@ -78,32 +78,31 @@ def validate_request(token, session, request, key, db, locale = nil)
...
-    if nonce[1] > Time.utc
-      db.exec("UPDATE nonces SET expire = $1 WHERE nonce = $2", Time.utc(1990, 1, 1), nonce[0])
-    else
-      raise translate(locale, "Erroneous token")
-    end
+  if !Crypto::Subtle.constant_time_compare(token["signature"].to_s, sign_token(key, token))
+    raise translate(locale, "Invalid signature")
   end
 
   scopes = token["scopes"].as_a.map { |v| v.as_s }
```
Comment (leonklingele) [inline src/invidious/helpers/tokens.cr]: The scope check could be made _after_ verifying the signature, what do you think @omarroth?

Comment (omarroth) [reply src/invidious/helpers/tokens.cr]: Makes sense to me.

**Comment by leonklingele** [comment]
Updated. ptal @omarroth

**Comment by omarroth** [comment]
LGTM 👍

**Pull request merged**